### ISSUE TITLE
GPU issues with NoGradCuArray

### ISSUE THREAD

**Issue by nmheim** [nmheim/GenerativeModels.jl#49]
```julia
using GenerativeModels
using Zygote
using Flux
using Debugger

enc = CMeanVarGaussian{Float32,DiagVar}(Dense(5,4))
dec = CMeanVarGaussian{Float32,ScalarVar}(Dense(2,6))
pri = Gaussian(ones(Float32, 2), ones(Float32, 2))

vae = VAE(pri, enc, dec)
x = ones(5)

vae = gpu(vae)
x = gpu(x)

loss() = elbo(vae, x)
ps = params(vae)

Zygote.pullback(loss, ps)
```
The above code works on the `gpu` branch, but as soon as there are `NoGrad(Cu)Array`s involved it breaks
```julia
using GenerativeModels
using Zygote
using Flux
using Debugger

enc = CMeanVarGaussian{Float32,DiagVar}(Dense(5,4))
dec = CMeanVarGaussian{Float32,ScalarVar}(Dense(2,6))
pri = Gaussian(NoGradArray(ones(Float32, 2)), ones(Float32, 2)) # <- NoGradArray

vae = VAE(pri, enc, dec)
x = ones(5)

vae = gpu(vae)
x = gpu(x)

loss() = elbo(vae, x)
ps = params(vae)

Zygote.pullback(loss, ps)
```

**Comment by nmheim** [comment]
Error message:
```julia
 ERROR: LoadError: GPU compilation of #25(CuArrays.CuKernelState, CUDAnative.CuDeviceArray{Tuple{Float32,Zygote.var"#1682#back#636"{Zygote.var"#634#635"}},2,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},Zygote.var"#1370#1377"{Zygote.Context,typeof(-)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},Zygote.var"#1370#1377"{Zygote.Context,typeof(-)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}.
That type is not isbits, and such arguments are only allowed when they are unused by the kernel.  .f is of type Zygote.var"#1370#1377"{Zygote.Context,typeof(-)} which is not isbits.
    .__context__ is of type Zygote.Context which is not isbits.
      .cache is of type Union{Nothing, IdDict{Any,Any}} which is not isbits.
      .globals is of type Union{Nothing, Dict{GlobalRef,Any}} which is not isbits.


Stacktrace:
 [1] check_invocation(::CUDAnative.CompilerJob, ::LLVM.Function) at /home/niklas/.julia/packages/CUDAnative/bMikt/src/compiler/validation.jl:70
 [2] macro expansion at /home/niklas/.julia/packages/CUDAnative/bMikt/src/compiler/driver.jl:187 [inlined]
 [3] macro expansion at /home/niklas/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
 [4] #codegen#140(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.codegen), ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/bMikt/src/compiler/driver.jl:186
 [5] #codegen at ./none:0 [inlined]
 [6] #compile#139(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.compile), ::Symbol, ::CUDAnative.CompilerJob) at /home/niklas/.julia/packages/CUDAnative/bMikt/src/compiler/driver.jl:47
 [7] #compile#138 at ./none:0 [inlined]
 [8] #compile at ./none:0 [inlined] (repeats 2 times)
 [9] macro expansion at /home/niklas/.julia/packages/CUDAnative/bMikt/src/execution.jl:400 [inlined]
 [10] #cufunction#182(::Nothing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::GPUArrays.var"#25#26", ::Type{Tuple{CuArrays.CuKernelState,CUDAnative.CuDeviceArray{Tuple{Float32,Zygote.var"#1682#back#636"{Zygote.var"#634#635"}},2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},Zygote.var"#1370#1377"{Zygote.Context,typeof(-)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}},Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,2,CUDAnative.AS.Global},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}) at /home/niklas/.julia/packages/CUDAnative/bMikt/src/execution.jl:368
 [11] cufunction(::Function, ::Type) at /home/niklas/.julia/packages/CUDAnative/bMikt/src/execution.jl:368
 [12] macro expansion at /home/niklas/.julia/packages/CUDAnative/bMikt/src/execution.jl:176 [inlined]
 [13] macro expansion at ./gcutils.jl:91 [inlined]
 [14] macro expansion at /home/niklas/.julia/packages/CUDAnative/bMikt/src/execution.jl:173 [inlined]
 [15] _gpu_call(::CuArrays.CuArrayBackend, ::Function, ::CuArray{Tuple{Float32,Zygote.var"#1682#back#636"{Zygote.var"#634#635"}},2}, ::Tuple{CuArray{Tuple{Float32,Zygote.var"#1682#back#636"{Zygote.var"#634#635"}},2},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},Zygote.var"#1370#1377"{Zygote.Context,typeof(-)},Tuple{Base.Broadcast.Extruded{CuArray{Float32,1},Tuple{Bool},Tuple{Int64}},Base.Broadcast.Extruded{CuArray{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at /home/niklas/.julia/packages/CuArrays/35T8v/src/gpuarray_interface.jl:60
 [16] gpu_call at /home/niklas/.julia/packages/GPUArrays/tIMl5/src/abstract_gpu_interface.jl:151 [inlined]
 [17] gpu_call at /home/niklas/.julia/packages/GPUArrays/tIMl5/src/abstract_gpu_interface.jl:128 [inlined]
 [18] copyto! at /home/niklas/.julia/packages/GPUArrays/tIMl5/src/broadcast.jl:48 [inlined]
 [19] copyto! at ./broadcast.jl:863 [inlined]
 [20] copy at ./broadcast.jl:839 [inlined]
 [21] materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.ArrayStyle{NoGradCuArray},Nothing,Zygote.var"#1370#1377"{Zygote.Context,typeof(-)},Tuple{CuArray{Float32,1},CuArray{Float32,2}}}) at ./broadcast.jl:819
 [22] _broadcast(::Zygote.var"#1370#1377"{Zygote.Context,typeof(-)}, ::NoGradCuArray{Float32,1}, ::Vararg{Any,N} where N) at /home/niklas/.julia/packages/Zygote/eFliU/src/lib/broadcast.jl:110
 [23] adjoint at /home/niklas/.julia/packages/Zygote/eFliU/src/lib/broadcast.jl:119 [inlined]
 [24] _pullback at /home/niklas/.julia/packages/ZygoteRules/Mmoki/src/adjoint.jl:47 [inlined]
 [25] adjoint at /home/niklas/.julia/packages/Zygote/eFliU/src/lib/lib.jl:139 [inlined]
 [26] _pullback at /home/niklas/.julia/packages/ZygoteRules/Mmoki/src/adjoint.jl:47 [inlined]
 [27] broadcasted at ./broadcast.jl:1237 [inlined]
 [28] _pullback(::Zygote.Context, ::typeof(Base.Broadcast.broadcasted), ::typeof(-), ::NoGradCuArray{Float32,1}, ::CuArray{Float32,2}) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface2.jl:0
 [29] kld at /home/niklas/repos/GenerativeModels.jl/src/pdfs/abstract_cgaussian.jl:34 [inlined]
 [30] _pullback(::Zygote.Context, ::typeof(kld), ::CMeanVarGaussian{Float32,DiagVar}, ::Gaussian{Float32}, ::CuArray{Float32,1}) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface2.jl:0
 [31] #elbo#19 at /home/niklas/repos/GenerativeModels.jl/test.jl:22 [inlined]
 [32] _pullback(::Zygote.Context, ::var"##elbo#19", ::Int64, ::typeof(elbo), ::VAE{Float32}, ::CuArray{Float32,1}) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface2.jl:0
 [33] elbo at /home/niklas/repos/GenerativeModels.jl/test.jl:20 [inlined]
 [34] _pullback(::Zygote.Context, ::typeof(elbo), ::VAE{Float32}, ::CuArray{Float32,1}) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface2.jl:0
 [35] loss at /home/niklas/repos/GenerativeModels.jl/test.jl:17 [inlined]
 [36] _pullback(::Zygote.Context, ::typeof(loss)) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface2.jl:0
 [37] pullback(::Function, ::Params) at /home/niklas/.julia/packages/Zygote/eFliU/src/compiler/interface.jl:96
 [38] top-level scope at /home/niklas/repos/GenerativeModels.jl/test.jl:20
 [39] include at ./boot.jl:328 [inlined]
 [40] include_relative(::Module, ::String) at ./loading.jl:1105
 [41] include(::Module, ::String) at ./Base.jl:31
 [42] include(::String) at ./client.jl:432
 [43] top-level scope at REPL[2]:1
in expression starting at /home/niklas/repos/GenerativeModels.jl/test.jl:20
```

**Comment by nmheim** [comment]
`NoGradCuArray`s are no longer used. closed by 8e277802c7aebd581b321e8fed24fb04914bb892
